pkg/varlog/operations.go

```diff
@@ -21,29 +22,59 @@ func (v *logImpl) append(ctx context.Context, tpid types.TopicID, lsid types.Log
 		opt.apply(&appendOpts)
 	}
 
+	lsidx := 0
+	var lsids []types.LogStreamID
+
+RETRY:
 	for i := 0; i < appendOpts.retryCount+1; i++ {
 		if appendOpts.selectLogStream {
-			var ok bool
-			if lsid, ok = v.lsSelector.Select(tpid); !ok {
-				err := fmt.Errorf("append: no usable log stream in topic %d", tpid)
-				result.Err = multierr.Append(result.Err, err)
-				continue
-			}
-			if _, ok = appendOpts.allowedLogStreams[lsid]; appendOpts.allowedLogStreams != nil && !ok {
-				err := fmt.Errorf("append: not allowed lsid %d", lsid)
-				result.Err = multierr.Append(result.Err, err)
+			if appendOpts.allowedLogStreams == nil {
+				var ok bool
+				if lsid, ok = v.lsSelector.Select(tpid); !ok {
+					err := fmt.Errorf("append: no usable log stream in topic %d", tpid)
+					result.Err = multierr.Append(result.Err, err)
+					continue RETRY
+				}
+			} else {
+				if len(lsids) == lsidx {
+					lsidx = 0
+					if lsids = v.lsSelector.GetAll(tpid); len(lsids) == 0 {
+						err := fmt.Errorf("append: no usable log stream in topic %d", tpid)
+						result.Err = multierr.Append(result.Err, err)
+						continue RETRY
+					}
+
+					rand.Shuffle(len(lsids), func(i, j int) { lsids[i], lsids[j] = lsids[j], lsids[i] })
+				}
+
+			SELECT:
+				for ; lsidx < len(lsids); lsidx++ {
+					lsid = lsids[lsidx]
+					if _, ok := appendOpts.allowedLogStreams[lsid]; !ok {
+						v.allowlist.Deny(tpid, lsid)
```

Comment:
Optional: I'm unsure whether it is good to put a log stream into denylist since it does not exist in the allowed log streams passed by users. Do you think this change is critical for this PR?

Reply:
It works without adding it to the deny list. However, there is a load of filtering every time. If the zombie LS are removed, the load of LS selection will be reduced. I'll delete it and add it if necessary.